mitiq/observable/pauli.py

```diff
@@ -161,3 +143,115 @@ def __str__(self) -> str:
+            if check_precondition and not self.can_add(pauli):
+                raise ValueError(
+                    f"Cannot add PauliString {pauli} to PauliStringSet."
+                )
+            weight = pauli.weight()
+            if self._paulis.get(weight) is None:
+                self._paulis[weight] = {pauli}
+            else:
+                self._paulis[weight].add(pauli)
+
+    @property
+    def elements(self) -> Set[PauliString]:
+        return {pauli for pset in self._paulis.values() for pauli in pset}
+
+    @property
+    def elements_by_weight(self) -> Dict[int, Set[PauliString]]:
+        return self._paulis
+
+    def support(self) -> Set[int]:
+        return {cast(cirq.LineQubit, q).x for q in self._qubits_to_measure()}
+
+    def weight(self) -> int:
+        return len(self.support())
+
+    def max_weight(self) -> int:
+        return min(self._paulis.keys())
+
+    def min_weight(self) -> int:
+        return max(self._paulis.keys())
+
+    def _qubits_to_measure(self) -> Set[cirq.Qid]:
+        qubits: Set[cirq.Qid] = set()
+        for pauli in self.elements:
+            qubits.update(pauli._pauli.qubits)
+        return qubits
+
+    def measure_in(self, circuit: QPROGRAM) -> QPROGRAM:
+        return self._measure_in(circuit, self)
+
+    @staticmethod
+    @atomic_converter
+    def _measure_in(
+        circuit: cirq.Circuit, pauliset: "PauliStringSet"
+    ) -> cirq.Circuit:
+        # Transform circuit to canonical qubit layout.
+        qubit_map = dict(
+            zip(
+                sorted(circuit.all_qubits()),
+                cirq.LineQubit.range(len(circuit.all_qubits())),
+            )
+        )
+        circuit = circuit.transform_qubits(lambda q: qubit_map[q])
+
+        if not pauliset._qubits_to_measure().issubset(
+            set(circuit.all_qubits())
+        ):
+            raise ValueError(
+                f"Qubit mismatch. The PauliString(s) act on qubits "
+                f"{pauliset.support()} but the circuit has qubit indices "
+                f"{sorted([q for q in circuit.all_qubits()])}."
+            )
+
+        basis_rotations = set()
+        support = set()
+        for pauli in pauliset.elements:
+            basis_rotations.update(pauli._basis_rotations())
+            support.update(pauli._qubits_to_measure())
+        measured = circuit + basis_rotations + cirq.measure(*sorted(support))
+
+        # Transform circuit back to original qubits.
+        reverse_qubit_map = dict(zip(qubit_map.values(), qubit_map.keys()))
+        return measured.transform_qubits(lambda q: reverse_qubit_map[q])
+
+    def _expectation_from_measurements(
+        self, measurements: MeasurementResult
+    ) -> float:
+        total = 0.0
+        for pauli in self.elements:
+            bitstrings = measurements[[q.x for q in pauli._pauli.qubits]]
+            value = np.average([(-1) ** np.sum(bits) for bits in bitstrings])
+            total += pauli.coeff * value
```

Comment:
Maybe it is more natural to move the body of this loop in the definition of `PauliString._expectation_from_measurements()`.   It would be more modular. 
But the current structure is also ok.

Reply:
I originally had it that way, but then I realized the more general way is to have it in `PauliStringCollection` since a `PauliString` is (trivially) a `PauliStringCollection`, but a `PauliStringCollection` is not a `PauliString`.